### ISSUE TITLE
Order for DependsOnGroups  has changed after TestNg 7.4.0

### ISSUE THREAD

**Issue by martinaldrin** [cbeust/testng#2664]
### TestNG Version
7.5 not working
7.4 works

### Expected behavior
s3
s1
s2

### Actual behavior

s1
s2
s3

### Is the issue reproducible on runner?

- [ ] Maven
- [ ] Eclipse
- [ ] IntelliJ

### Test case sample

Here is an example for BeforeSuite, but the same problem exists for all configuration methods. The same issue also for dependsOnMethod

```
    @BeforeSuite(groups = { "g1" })
    public void s3() {
        System.out.println("s3");
    }

    @BeforeSuite(groups = { "g2" }, dependsOnGroups = "g1")
    public void s2() {
        System.out.println("s2");
    }

    @BeforeSuite(groups = { "g2" }, dependsOnGroups = "g1")
    public void s1() {
        System.out.println("s1");
    }

    @Test()
    public void test3() {
        System.out.println("test3");
    }
```

I have tried to find the commit where the fault was introduced, but my IDE did not like to switch gradle version and I'm new at gradle.

**Comment by martinaldrin** [comment]
As I understand it When we add back the sorting based on groups we get cyclic dependencies. If the methods are spread out to different classes.

And as I understand it the problem is this method, where a a method dependency is added to solve issues that methods in base classes are executed before the methods in the test classes. And If we add group dependsOn we get duplicated dependencies which give the cyclic dependencies.
MethodInheritance.fixMethodInheritance()
   m2.addMethodDependedUpon(MethodHelper.calculateMethodCanonicalName(m1));

**Comment by juherr** [comment]
@martinaldrin I think you got some part of the design issue.
You can add test cases in a PR if you have some in mind ;)

**Comment by martinaldrin** [comment]
Hi @juherr, Yes I understand part of the problem now. but I have no clue about how to solve it and I guess I will need your expert help to solve it. Yes I will add more unit tests to my PR.

**Comment by martinaldrin** [comment]
But after more digging. I actually think the behavior that we try to protect in test.configuration.issue2432.IssueTest is wrong.
Test expect following behaviour:
```
"prepareConfig",
"uploadConfigToDatabase",
"verifyConfigurationAfterInstall",
"prepareConfigForTest1",
"test1"
```
I belive that following behaviour is correct:
```
"prepareConfig",
"prepareConfigForTest1",
"uploadConfigToDatabase",
"verifyConfigurationAfterInstall",
"test1"
```

both prepareConfig and prepareConfigForTest1 have the group "prepareConfig" and therefor I they should be executed first regardless of which class they are in. If we don't want that that behavior I think the groups between the classes should be different.  otherwise it will not be possible to define groups into different classes.

**Comment by martinaldrin** [comment]
I created a separate PR where I use my priority implementation to handle the class hierarchy sorting.  
But unit test that I showed above needs to be changed, since that will not be possible to solve, without breaking other things.

https://github.com/cbeust/testng/pull/2667

**Comment by martinaldrin** [comment]
One more PR, showing an alternative solution where I use a different priority to not interfere with the user configurations. 

https://github.com/cbeust/testng/pull/2668

**Comment by juherr** [comment]
@martinaldrin Thanks for all your contributions.

> I guess I will need your expert help to solve it.

To be honest, I've no idea how to deal with it because, except for selection, the groups feature is not yet clearly designed and may be difficult to integrate into the current base code.

**Comment by martinaldrin** [comment]
@juherr I think it works pretty well for everything except for one that was recently implemented after 7.4 release. 
With my changes I think groups starts to work as it did before
